### ISSUE TITLE
Origins

### ISSUE THREAD

**Issue by Virtlink** [Virtlink/redgreen-aterms#5]
Spoofax uses _origin tracking_ to track where terms came from. This information must be added to the green tree, and be kept under transformations.
- [ ] Add nullable origin information to proto-terms.
- [ ] Add an option to copy origin information to a replacement term (in `withChild`).
- [ ] Write example for working with origin information.
- [ ] Write test for origin information.

**Comment by Apanatshka** [comment]
Doesn't adding this to green tree make all green nodes unique and unshareable in normal parse trees? Is the width info not enough to reconstruct origin info in the red trees?

**Comment by Virtlink** [comment]
Yes and yes. Adding origins will makes each node unique and therefore unshareable, and the width info should be enough to reconstruct the origin info.

However, once we start moving subtrees and terms around (term rewriting, transformations, desugarings), the only way to know where a term came from (e.g. for error reporting, highlighting, etc) is by storing the origin information explicitly with the term.

**Comment by Apanatshka** [comment]
I can think of other ways.. Like a transformed term points back to the origin parse tree, where it came from. Of course this might be problematic with the original parse tree needing to stay alive all the time and further and further away from the information source that you derive origin info from..
